Process and Mappings Cleanup with Resource Reporting

gmarler opened the pull request.

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+    fn report_resource_consumption(&mut self) {
+        let (mut exited_procs, mut running_procs) = (0, 0);
+        let procs_guard = self.procs.read();
+        for (_pid, proc_info) in procs_guard.iter() {
```
nit
```suggestion
        for proc_info in procs_guard.values() {
```

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+        let (mut exited_procs, mut running_procs) = (0, 0);
+        let procs_guard = self.procs.read();
+        for (_pid, proc_info) in procs_guard.iter() {
+            if proc_info.status == ProcessStatus::Exited {
```
nit: Use match to ensure we cover all variants

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+            "{} process exits this session; {} process exits since profiler startup",
+            self.exit_proc_per_session, self.exit_proc_total
+        );
+        info!(
```
IMO this should be a `debug` log

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+                .info()
+                .unwrap()
+                .info
+                .max_entries
```
This number is 'static' once the BPF map is created, right?

gmarler replied:
So, grab it once and use it forever?  If they ever make map's adjustable in size, I can imagine that could pose an issue.

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+                .maps
+                .exec_mappings
+                .info()
+                .unwrap()
```
If possible, let's avoid adding new unwraps to the code. I am aware the codebase is littered with some uncessary unwraps which I am planning to tackle soon. If you think it's an invariant that should always hold true, `expect("$invariant")` is preferred in this project

javierhonduco commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+        // work after a couple of sessions have elapsed.
+        self.deletion_scheduler
+            .write()
+            .add(ToDelete::Process(Instant::now(), pid, partial_write));
```
Is the timestamp correct? Need to check the API because I don't remember atm, but leaving a note so I don't forget

gmarler replied:
It doesn't have to be _that_ correct, as the DeletionScheduler makes PIDs "pending" for deletion after 2 sessions worth of time (10 seconds at the moment), so the resolution required is much more than a second.

gmarler replied:
So, you want to store the _future_ Instant beyond which the PID deletion can occur, instead of the Instant at which the PID was added to the Deletion Scheduler?

Either works, of course, and I didn't see the API spelled out, so I went with what seemed intuitive.  Glad to pick one way or the other, as long as it doesn't prolong this PR merging.

javierhonduco commented on src/profiler.rs:
```diff
@@ -1712,8 +1785,15 @@ impl Profiler {
+                // Could be other things too, like eviction failures
+                match e {
+                    AddProcessError::ProcfsRace => (),
+                    _ => {
```
style nit: prefer `if let` to deconstruct this enum if we want to only discrimiate against one variant (or in this case, everything but one variant)

gmarler replied:
I'm confident the way this works is correct, but need to discuss how to add a test so that if I make this an `if let...` I still have confidence that it's working.

javierhonduco commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+                .deletion_scheduler
+                .write()
+                .pop_pending(pending_duration);
+            if !to_delete_vec.is_empty() {
```
Two things here:

- Since you are checking if the vector isn't empty and then fetching the first element, why not match against it? The code would be cleaner that way
- Since `pop_pending` returns a vector, aren't we potentially not schuling pids `1..n`?

javierhonduco replied:
Yeah that's very confusing, given that `pop_pending` returns a single element, it should definitely not be a vector. I should have caught that during code review!

gmarler replied:
All fixed now.

javierhonduco commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+                        // - How many mappings per PID (mappings_by_pid HashMap)
+                        let mut mappings_by_pid: HashMap<u32, u32> = HashMap::new();
+                        // - Mappings we're going to have to manually delete
+                        let mut mappings_to_delete: Vec<_> = vec![];
```
nit:
```suggestion
                        let mut mappings_to_delete = vec![];
```

javierhonduco commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+                            );
+                        }
+                        // Now, delete those mappings
+                        for key in mappings_to_delete.iter() {
```
Isn't this logic duplicated elsewhere? If this code needs to run perhaps it would be a symptom of a bigger issue that would be worth tackling.

gmarler replied:
I'll look at that, it's certainly possible I moved it from elsewhere and forgot to remove the source, which might account for one issue I'm seeing on rare occasions - certain mappings are already gone when we get around to deleting them.  Just note those and move on.

gmarler commented on src/bpf/tracers.bpf.c:
```diff
@@ -49,9 +49,12 @@ int tracer_process_exit(void *ctx) {
+    // To avoid a race condition with detecting and registering a PID *before*
+    // its exit tracepoint has fired, we just ignore whether we know about this
+    // PID or not, at least inside this BPF program
+    // if (!process_is_known(per_process_id)) {
```
Fine by me

javierhonduco replied:
Let's think a bit more on the trade-offs of sending the event even if the process does not exist, as this will incur in more work for userspace

gmarler replied:
Measured impact via all of:
* a `span!` (which I've since removed) in `handle_process_exit()` to see how high that latency can go
* The 997Hz profile that barely registers `handle_process_exit()` running at all - ran that on a few other machines other than the host with high short-lived process spikes
* And running on several workloads with very high process counts (tens of thousands), as well as very high short-lived process spikes, outside our test machine.

gmarler commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+            .native_unwinder
+            .maps
+            .exec_mappings
+            .keys()
```
For now it's critical to track down what's going on with these mappings and should be on by default in all cases until we eliminate all of the corner cases.

It's also only done once a session, and over 5000 PIDs, so it should be _that bad_.

But I'm more than happy to wrap this in a timed log message so we can see if it's truly becoming a problem everywhere.

I do fully expect to remove this completely as we eliminate any final issues around mappings.

gmarler replied:
Yah, I'll get that done today most likely.  But at this point performance takes a back seat to correctness.  The current git repo's code is demonstrably incorrect, as it runs out of memory in minutes/hours, and we're peeling the onion to get to a state where we can run lightswitch for weeks without worrying about that.

Being overly concerned about the performance at this stage seems to be premature optimization.

However, having the profile for future reference when making future changes makes good sense, especially as we improve _how_ this PR's code does what it does over time.

gmarler replied:
confirmed - 997Hz profile doesn't show `report_resource_consumption()` at all with the most recent changes made in this PR.

gmarler commented on src/profiler.rs:
```diff
@@ -2089,6 +2363,24 @@ impl Profiler {
     }
+
+    fn live_pid_count(&mut self) -> usize {
+        let live_pids: Vec<Pid> = procfs::process::all_processes()
```
Although we don't have any machines with more than a few 10s of thousands of processes (and less than 10 of those, if that) - others might though.

I'm not against gating this with an option, but we need this information now for all hosts as we go through beta, without thinking about turning it on - it just needs to be there.  Definitely agree on adding the option in 2026 though, as we get more stability overall.

Not seeing it be that expensive either at present - either in terms of CPU or memory.

javierhonduco replied:
I am fine with adding this since it can be useful but definitely this should be opt-in. 

> Not seeing it be that expensive either at present - either in terms of CPU or memory.

With 10ks of processes this might be fast enough, but on a machine with 10x or 100x that number and under severe resource pressure this might become a hotspot. 

Either way, interested in seeing the flamegraph (at 997 Hz plz) for the machine you are testing this on. Would be good to see what's the performance cost of the current code for your workloads you are testing it on.

gmarler replied:
I can also add a `span!` for this, but also note that this isn't showing up at all in the 997Hz profile, so that's good.  Gating this with an option in a subsequent PR if it becomes a problem anywhere is very palatable as well.

gmarler commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+            // 2nd pass - Perform the actual final deletions
+            let pids_to_del_str = pending_deletion
+                .iter()
+                .map(|(n, _)| n.to_string())
```
I'll give that a whirl

gmarler commented on src/profiler.rs:
```diff
@@ -852,41 +870,96 @@ impl Profiler {
+            .maps
+            .exec_mappings
+            .keys()
+            .map(|key| exec_mappings_key::from_bytes(&key).unwrap().pid)
```
I think I've got the way to `filter_map()` the Results of the `exec_mappings_key::from_bytes()` calls, while logging errors, and only keeping valid results.  See if you like it when I push the next commit.

javierhonduco commented on src/profiler.rs:
```diff
@@ -2089,6 +2363,24 @@ impl Profiler {
+
+    fn live_pid_count(&mut self) -> usize {
+        let live_pids: Vec<Pid> = procfs::process::all_processes()
+            .expect("Cannot read proc")
```
Sure, happy to help!

javierhonduco commented on src/profiler.rs:
```diff
@@ -2165,19 +2484,19 @@ mod tests {
-                .count(),
-            0
+                .count()
+                > 0
```
(I thought I had left a note here but perhaps I removed it by accident?)

This test needs an update to ensure that your changes are working properly and that this code doesn't break by accident. 

A possible approach would be to manually trigger a deletion after these asserts, and then assert that all the releveant data structures are, as expected, deleted.

gmarler replied:
Will definitely need either an example/detailed description of how to express such a test, or to discuss over Zoom via pairing.

gmarler commented on src/profiler.rs:
```diff
@@ -1788,7 +1868,11 @@ impl Profiler {
     }
 
     pub fn add_proc(&mut self, pid: Pid) -> Result<(), AddProcessError> {
+        // NOTE: There are 3 places where AddProcessError::ProcfsRace can be returned from this
```
I tend to document ONLY in code, never in a PR - because I've never gone back and re-read a PR - not seeing the benefit as yet.

gmarler commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+        // TODO: Wrap this function's activities in a logged duration timer
+        // Pop off any processes that we've kept around long enough after they've exited
+        // Where "long enough" is 2 sessions worth
+        let pending_duration = self.session_duration.checked_mul(2).unwrap();
```
You don't have an option, from what I can see - `checked_mul()` is the only way to do a multiplication, and I want the code to be explicit about what and why we're doing this action.

And since both `session_duration` and 2 are constants (once the program is running), there's never a chance of a panic.

gmarler replied:
Fixed this.

javierhonduco commented on src/bpf/tracers.bpf.c:
```diff
@@ -49,9 +49,12 @@ int tracer_process_exit(void *ctx) {
-    if (!process_is_known(per_process_id)) {
-        return 0;
-    }
+    // To avoid a race condition with detecting and registering a PID *before*
```
Can you please clarify where you saw these logs? The path of the file or the command line you ran to see them would really help to understand what's really going on

gmarler commented:
Ok, here's the 60 second profile @997Hz of lightswitch built from this PRs current state, taken by another instance of lightswitch, but limiting to just to the originally running lightswitch (there's got to be a better way to explain that):

![flame](https://github.com/user-attachments/assets/d04942c0-b640-4930-9469-429ac0cbb643)

This machine has a very small workload, but high short-lived process rates, as shown by the following final message over the 60 second period - had to look very closely to see any usage of `handle_process_exit()`  But obviously seeing lots of work in `cleanup_procs()`, but related to iterating over the `exec_mappings` map, as one might expect, which only happens for PIDs where we actually had mappings to begin with.

```
INFO report_resource_consumption: lightswitch::profiler: 4861 process exits this session; 32066 process exits since profiler startup
```

javierhonduco commented on src/profiler.rs:
```diff
@@ -2020,6 +2132,195 @@ impl Profiler {
+        // Where "long enough" is 2 sessions worth
+        let pending_duration = self
+            .session_duration
+            .checked_mul(2)
```
We agreed that if this can't overflow, there's no need for a checked multiplication, no?

javierhonduco commented on src/profiler.rs:
```diff
@@ -2020,6 +2132,195 @@ impl Profiler {
     }
 
+    fn cleanup_procs(&mut self) {
+        // Wrap this function's activities in a logged duration timer
```
nit: This comment is not totally correct, the function is not being wrapped. I would see that the comment is not needed as the behaviour of `span!` should be well understood for engineers perusing this codebase. We can always check the documentation if we have doubts

gmarler commented on src/profiler.rs:
```diff
@@ -2020,6 +2112,188 @@ impl Profiler {
+                };
+                pending_deletion.push((pid, partial_write));
+            } else {
+                break;
```
Yep, and this is working as expected now.

gmarler commented:
Made modifications to `cleanup_procs()` to only iterate through `exec_mappings` once per session, rather than once per dead PID mapping per session.

The 997 Hz profile looks quite a bit better:
![bbperf-tt-774-flame-20251224](https://github.com/user-attachments/assets/55ef529f-9913-42b8-a4f2-194105b5d115)

* `cleanup_procs()` looks to have been optimized away
* but I can see things it calls, like `pop_pending()` - just barely

gmarler commented:
And that's what it was.

@javierhonduco, I think you can re-review `cleanup_procs()` in particular, and this PR in general, because after the recent changes based on your first review:

* I barely see `cleanup_procs()` function at all in the 997 Hz profile (see the last profile above - still looks the same after all of the commits between then and this update).  Don't see `report_resource_consumption()` in the profile at all.
* Both procs and mappings are completely cleaned up with no leaks of mappings after many hours of operation after a couple of hiccups from my most recent changes - all fixed now
* I think I've cleaned up both all the issues you highlighted, and the nits - thanks much for those!
* Added span timing for both `cleanup_procs()` and `report_resource_consumption()` along the way, so we can explicitly know how much CPU they're actually consuming.  The former no longer gets anywhere close to using more than 100 ms per session (it was using up to 6 seconds (!!!) with the previous iteration pattern in the worst cases).  The latter no longer exceeds 18 ms per session.

gmarler reviewed (commented):
I think I've identified two places where mappings are being lost on process exit, and they're being reported by my last debug change.

After running for several hours, all lost mappings that need cleaning up are caused by one or the other.  So now we need to discuss how to prevent these corner cases, or at least clean up after them efficiently.
